### ISSUE TITLE
Performance of search Weblate's Translation Memory 

### ISSUE THREAD

**Issue by pavelhoral** [WeblateOrg/weblate#9289]
### Describe the issue

PostgreSQL query planner can opt for using sequential scan when searching for TM suggestions. This leads to degraded performance when TM contains hundreds of thousands entries. This behavior can be avoided if the TM query contains ORDER BY clause that actually forces PostgreSQL to use index scan.

<details><summary>Example query that is being generated right now:</summary>

```sql

DECLARE "_django_curs_140138876151616_sync_430" NO SCROLL CURSOR WITH HOLD FOR
SELECT
  "memory_memory"."id",
  "memory_memory"."source_language_id",
  "memory_memory"."target_language_id",
  "memory_memory"."source",
  "memory_memory"."target",
  "memory_memory"."origin",
  "memory_memory"."user_id",
  "memory_memory"."project_id",
  "memory_memory"."from_file",
  "memory_memory"."shared"
FROM "memory_memory"
WHERE
  (
    (
      "memory_memory"."from_file" OR
      "memory_memory"."project_id" = 2 OR
      "memory_memory"."user_id" = 6
    ) AND
    "memory_memory"."source" % 'Lorem Ipsum' = true AND
    "memory_memory"."source_language_id" = 148 AND
    "memory_memory"."target_language_id" = 114
  )
LIMIT 50;
```
</details>

Query mentioned above should have `ORDER BY similarity(source, 'Lorem Ipsum')` added to make sure index is actually used. Otherwise the behavior of PostgreSQL is sort of unpredictable.

---

Reference to my investigation - https://github.com/WeblateOrg/weblate/discussions/9284

I would gladly try to fix this myself, but I am not sure where to start to not break more stuff.

### I already tried

- [X] I've read and searched [the documentation](https://docs.weblate.org/).
- [X] I've searched for similar issues in this repository.

### Steps to reproduce the behavior

* Create translation project with a large translation memory.
* Start the translation and observe speed of `Automatic suggestion` API.

* Alternativelly simply try doing `EXPLAIN ANALYZE` on the query that is being used by the backend.
* Quite often PostgreSQL uses sequential scan for that query which can take multiple seconds to execute.

### Expected behavior

Queries generated when fetching automatic suggestions are always using databse indexes and return entries in a matter of milliseconds or just a few hundreds of milliseconds at most.

### How do you run Weblate?

Docker container

### Weblate versions

* Weblate Docker image 4.17
* PostgreSQL Docker image 15.2-bullseyye

**Comment by tomkolp** [comment]
In my project, when the number of source strings exceeded 500,000, then in one component the autosuggestions stopped working completely.  The problem has been around for 2 years.  The source strings in this file have a lot of words, because this is the content of the books, and we can't split it, because the translator will lose the context.  The auto-suggest query continues until the request timeout expires and the search ends.  In other components the search for autosuggestions is acceptable (5 to 20 seconds) My project is gigantic in size.
https://weblate.eso-spolszczenie.eu/projects/eso-spolszczenie/21337012/pl/
![Screenshot_20230522_204541_Edge](https://github.com/WeblateOrg/weblate/assets/39633242/58e0f0be-dbfe-4949-9021-dab511b4ddd3)

**Comment by pavelhoral** [comment]
I am still trying to understand bit more what is going on under the hood. I might try to create some simple replicable test case only with PostgreSQL. I don't have much experience with pg_trgm. The issue with PostgreSQL not using indexes is probably known judging on this comment - https://github.com/WeblateOrg/weblate/blob/main/weblate/utils/db.py#L35-L36.

**Comment by nijel** [comment]
Will removing CURSOR help as well? It really is not needed here, so I've removed that in 959f30380c83d33ceb953f209afb307ec38359f2.

Also, will really `ORDER BY similarity(source, 'Lorem Ipsum')` work? Using functions for lookups didn't utilize index, that's why we adjust `pg_trgm.similarity_threshold` and use `%` operator instead.

**Comment by nijel** [comment]
https://github.com/WeblateOrg/weblate/pull/9292 adds ordering, it would be great if you can test it on your data.

**Comment by pavelhoral** [comment]
> Also, will really ORDER BY similarity(source, 'Lorem Ipsum') work?

I use ORDER BY to make sure certain database indexes are preferred by query planner quite often. Although I know how and why that works only for B-Tree indexes. I am not sure if the same reasoning is still valid for other types of indexes (GIN / GiST)... probably not.

I will try to test the changes from PR, but first I want to learn more how pg_trig works, mainly because of https://github.com/WeblateOrg/weblate/discussions/9284#discussioncomment-5974944. Seems like using the index can still lead to quite unacceptable query times.

**Comment by pavelhoral** [comment]
Ok, I went through [GIN documentation](https://www.postgresql.org/docs/current/gin.html) and the ORDER BY is nonsense in this case. It worked when I have played with it but that was only a coincidental sideeffect. So my apologies for an incorrect assumption.

I am still trying to learn if there is anything else that can be done here.

**Comment by pavelhoral** [comment]
So to sum this up. Everything is as it is supposed to be... **without the cursor** everything works and the stuff I was missing in my tests was actually setting similarity threshold (default is 0.3 that matches too many tuples that needs to be rechecked). With the thershold [used by Weblate](https://github.com/WeblateOrg/weblate/blob/d3d4a98f49a5b204075e729087854399dd9537fb/weblate/machinery/weblatetm.py#L46) (hope it is this value) everything is blazingly fast (at least in my case).

Because the cursor is removed I think this issue can be closed as solved.

**Comment by tomkolp** [comment]
I just upgraded to:
4.18-dev — d3d4a98f49a5b204075e729087854399dd9537fb
Prior to the update in the books component, I was waiting 30 minutes for "automatic suggestions" and the request timed out.  After the update I wait 60-90s, it's still a long time, but it's much faster.

**Comment by nijel** [comment]
> So to sum this up. Everything is as it is supposed to be... **without the cursor** everything works and the stuff I was missing in my tests was actually setting similarity threshold (default is 0.3 that matches too many tuples that needs to be rechecked). With the thershold [used by Weblate](https://github.com/WeblateOrg/weblate/blob/d3d4a98f49a5b204075e729087854399dd9537fb/weblate/machinery/weblatetm.py#L46) (hope it is this value) everything is blazingly fast (at least in my case).

0.95 is used for strings matching, the translation memory uses something between 0.5 and 1 depending on the string length:

https://github.com/WeblateOrg/weblate/blob/d3d4a98f49a5b204075e729087854399dd9537fb/weblate/memory/models.py#L65-L69

**Comment by pavelhoral** [comment]
I would like to reiterate where the performance is being hit the most:

* GIN matches all **indexed items** based on [trigram similarity](https://github.com/postgres/postgres/blob/a817edbf6f302c376f5c0012d19a0474b6bdea88/contrib/pg_trgm/trgm_gin.c#L170) with the source string from the Weblate query:
   * This matches all items that contain **up to "similarity coefficient"** trigrams.
   * Matched items can be significantly longer than the queried string - this is how GIN works (the only condition is that those items have to **contain** what is being queried).
     [Actual match is rechecked later.](https://github.com/postgres/postgres/blob/a817edbf6f302c376f5c0012d19a0474b6bdea88/contrib/pg_trgm/trgm_gin.c#L185)
   * So if there are 1000 items in the form `Hello World, <random words from this point on>` and the query is `source ~ 'Hello World'`, GIN will match all 1000 items that have to be rechecked after fetching the tuples.
* After matching indexed entries GIN actually has to fetch tuples from the disk (pages / buffers) and recheck the condition against actual data.
   * This is where pg_trgm can say, that `Hello World, <random words>` does not actually match `Hello Word` because it contains a lot more trigrams.
   * **Fetching data from the disk only to discard them as not matching is what is the performance bottleneck.**
   * But even recomputing from shared buffers to simply throw all results away can be quite expensive (depends on how [different in terms of length](https://github.com/postgres/postgres/blob/a817edbf6f302c376f5c0012d19a0474b6bdea88/contrib/pg_trgm/trgm_gin.c#L302) those items are).

The two-stage process can be seen in the example query plan:

<details><summary>EXAMPLE QUERY PLAN</summary>

First index scan matches 156 index entries which are then rechecked and all are thrown away as not matching.

```
weblate=> EXPLAIN (ANALYZE, BUFFERS) SELECT * FROM "memory_memory"
WHERE "memory_memory"."source" % 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.' LIMIT 50;
                                                               QUERY PLAN                                                                
-----------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1236.12..1297.62 rows=16 width=205) (actual time=97.629..97.630 rows=0 loops=1)
   Buffers: shared hit=411 read=544
   ->  Bitmap Heap Scan on memory_memory  (cost=1236.12..1297.62 rows=16 width=205) (actual time=97.627..97.628 rows=0 loops=1)
         Recheck Cond: (source % 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.'::text)
         Rows Removed by Index Recheck: 156
         Heap Blocks: exact=82
         Buffers: shared hit=411 read=544
         ->  Bitmap Index Scan on memory_source_trgm  (cost=0.00..1236.12 rows=16 width=0) (actual time=41.962..41.963 rows=156 loops=1)
               Index Cond: (source % 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.'::text)
               Buffers: shared hit=238 read=400
 Planning:
   Buffers: shared hit=239 read=8
 Planning Time: 4.923 ms
 Execution Time: 98.040 ms
```
</details>


Based on the observation above the main issues are the following:

* If the similarity boundary is too low and the initial GIN match returns too many matches to be thrown away, the disk IO will slow down things significantly.
* If the searched value is too short and is contained in a lot of documents, GIN will match too many documents that will be thrown away in the recheck stage.
  * This means that having low `pg_trgm.similarity_threshold` even for short strings brings significant performance penalty as it matches more indexed items.


There are two things that can be done:

* There is [`gin_fuzzy_search_limmit` setting](https://www.postgresql.org/docs/current/gin-tips.html) that can limit the number of matched items during the first stage (if I understand that correctly).
   * This can help limitting number of matched items with the downside of actually not potentially finding the correct results.
   * This is database wide configuration setting and I don't think it is a good approach (compared to the next point).
* The pg_trgm index tuple can be extended with the actual string length which will prevent matching strings with significantly different length.
   * If I understand GIN and pg_trgm correctly, the actual length difference is tested only during the recheck phase after tuples are read.
     This can be prevented with index `USING GIN (source gin_trgm_ops, LENGTH(source))` and queries like `source % 'Lorem ipsum' AND LENGTH(source) BETWEEN 20 AND 30`.
     

<details><summary>Using LENGTH as part of the query:</summary>

```
weblate=> CREATE INDEX test_idx ON memory_memory USING GIN (source gin_trgm_ops, LENGTH(source));
CREATE INDEX

weblate=> EXPLAIN (ANALYZE, BUFFERS) SELECT * FROM memory_memory 
WHERE source % 'Lorem ipsum' AND LENGTH(source) BETWEEN 20 AND 30 LIMIT 50;
                                                         QUERY PLAN                                                          
-----------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=404.00..408.02 rows=1 width=205) (actual time=18.509..18.510 rows=0 loops=1)
   Buffers: shared hit=196
   ->  Bitmap Heap Scan on memory_memory  (cost=404.00..408.02 rows=1 width=205) (actual time=18.508..18.508 rows=0 loops=1)
         Recheck Cond: ((source % 'Lorem ipsum'::text) AND (length(source) >= 20) AND (length(source) <= 30))
         Buffers: shared hit=196
         ->  Bitmap Index Scan on test_idx  (cost=0.00..404.00 rows=1 width=0) (actual time=18.505..18.506 rows=0 loops=1)
               Index Cond: ((source % 'Lorem ipsum'::text) AND (length(source) >= 20) AND (length(source) <= 30))
               Buffers: shared hit=196
 Planning:
   Buffers: shared hit=2
 Planning Time: 2.564 ms
 Execution Time: 18.547 ms
(12 rows)
```
</details>

**I think that the length approach is an interesting one and should be tested.**

**Comment by nijel** [comment]
The length limitation makes the situation worse for me:

```
weblate=# EXPLAIN (ANALYZE, BUFFERS) SELECT * FROM memory_memory 
WHERE source % 'Lorem ipsum'  LIMIT 50;
                                                                 QUERY PLAN                                                                 
--------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=271.94..327.15 rows=50 width=148) (actual time=27082.165..27101.037 rows=50 loops=1)
   Buffers: shared hit=10881 read=186
   ->  Bitmap Heap Scan on memory_memory  (cost=271.94..3753.50 rows=3153 width=148) (actual time=27082.162..27101.020 rows=50 loops=1)
         Recheck Cond: (source % 'Lorem ipsum'::text)
         Rows Removed by Index Recheck: 329
         Heap Blocks: exact=228
         Buffers: shared hit=10881 read=186
         ->  Bitmap Index Scan on memory_source_trgm  (cost=0.00..271.15 rows=3153 width=0) (actual time=157.850..157.851 rows=989 loops=1)
               Index Cond: (source % 'Lorem ipsum'::text)
               Buffers: shared hit=5642 read=156
 Planning:
   Buffers: shared hit=1
 Planning Time: 1.357 ms
 Execution Time: 27101.103 ms
(14 rows)
```


```
weblate=# EXPLAIN (ANALYZE, BUFFERS) SELECT * FROM memory_memory 
WHERE source % 'Lorem ipsum' AND LENGTH(source) BETWEEN 20 AND 30 LIMIT 50;
                                                             QUERY PLAN                                                             
------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=923.10..941.05 rows=16 width=148) (actual time=43665.324..43665.326 rows=0 loops=1)
   Buffers: shared hit=20367
   ->  Bitmap Heap Scan on memory_memory  (cost=923.10..941.05 rows=16 width=148) (actual time=43665.322..43665.323 rows=0 loops=1)
         Recheck Cond: ((source % 'Lorem ipsum'::text) AND (length(source) >= 20) AND (length(source) <= 30))
         Rows Removed by Index Recheck: 796
         Heap Blocks: exact=462
         Buffers: shared hit=20367
         ->  Bitmap Index Scan on test_idx  (cost=0.00..923.10 rows=16 width=0) (actual time=5243.649..5243.650 rows=796 loops=1)
               Index Cond: ((source % 'Lorem ipsum'::text) AND (length(source) >= 20) AND (length(source) <= 30))
               Buffers: shared hit=13210
 Planning:
   Buffers: shared hit=2
 Planning Time: 2.499 ms
 Execution Time: 43700.125 ms
(14 rows)
```

This is on PostgreSQL 13, I didn't yet find time to upgrade this server.

**Comment by pavelhoral** [comment]
The query with LENGTH clause is using invalid length limit that is "out of bounds" for `Lorem Ipsum` and that is a mistake on my part when I used those numbers in my example above (I mainly wanted to demonstrate that the length is being evaluated during the inner most index scan).

So in your sample without the LENGTH clause the query returns 50 rows while the one with LENGTH returns 0 rows because the length effectively filters tuples that are being rejected during GIN recheck.

### TL;DR - this is the real deal, i promise :)

Create new index:

```sql
CREATE INDEX real_deal_index ON memory_memory USING GIN (
  source gin_trgm_ops,
  target_language_id,
  source_language_id,
  array_length(show_trgm(source), 1)
);
```

And use this to actually compare performance:

```sql
EXPLAIN (ANALYZE, BUFFERS)
SELECT * FROM memory_memory
WHERE
  source % 'Lorem ipsum' AND
  array_length(show_trgm(source), 1) <= (array_length(show_trgm('Lorem ipsum'), 1) / show_limit())
LIMIT 50;
```

I know this looks stupidly complex and is probably not something you would want to implement, but this will make the worst case scenario better IMO. So even if this is something that will never be implemented, it can be used to know what can be achieved in an ideal situation.

---

### Bit more detailed reasoning

The actual length boundaries (i.e. arguments to `BETWEEN ? AND ?`) should be in line with what the pg_trgm's similarity operator considers as similar documents.

Thinking about this again, we can ditch the lower boundary as that won't get past the first index search as it will already fail the first similarity check. So no need for `BETWEEN` only the upper boundary is needed. 

Next, looking at the pg_trgm source code:

* this is function to get similarity number based on trigrams of the query filter and the matched item - https://github.com/postgres/postgres/blob/a817edbf6f302c376f5c0012d19a0474b6bdea88/contrib/pg_trgm/trgm_op.c#L998
* this is the actual similarity calculation method - https://github.com/postgres/postgres/blob/ad5406246bff5490fbd428c148717c5b87e52ab1/contrib/pg_trgm/trgm.h#L112

The actual similarity forumla is (as linked above):

$$\text{similarity(a, b)}= \frac{count_{\text{shared trigrams}}}{length_{\text{a in trigrams}}  + length_{\text{b in trigrams}} - count_{\text{shared trigrams}}}$$

The worst case is if the matched document matches everything (i.e. $count_{\text{shared trigrams}} = length_\text{filter in trigrams}$) and then have some additional trigrams that will lower the similarity score. In that case the following inequality must hold:

$$\frac{length_{\text{filter in trigrams}}}{threshold} \geq length_{\text{matched item in trigrams}}$$

All the equation above are using trigrams... and the unfortunate truth is that there is no straightforward relation between number of trigrams in a string and length of that string. You can have string of one million spaces with a single "HELLO WORLD" at the end and even though length of the matched string is over one million, there are exactly 12 trigrams in it. So using LENGTH to limit the number of matched elements is strictly speaking not correct.

There is a "code smell" in my suggested solution though - function `show_trgm` that I am using to get the number of trigrams in a string [has the following description](https://www.postgresql.org/docs/current/pgtrgm.html#id-1.11.7.44.6):

> Returns an array of all the trigrams in the given string. (**In practice this is seldom useful except for debugging**.)

It might make sense to confirm this approach with PostgreSQL maintainers or pg_trgm authors before actually implementing it. There is also a slight possibility that what I wrote here is not 100 % correct.

**Comment by nijel** [comment]
Thanks for the research! Another thing which might be worth investigating for big translation memories is partitioning.

**Comment by pavelhoral** [comment]
One of PostgreSQL contributors replied to my original question regarding pg_trgm performance and it looks very relevant here - https://www.postgresql.org/message-id/CAMkU%3D1x2VvNSH_wHprQhffjq9CQJmuADBCsXhn7d05v_jUXMzg%40mail.gmail.com .

**Comment by nijel** [comment]
Have you done some benchmarks if adding length to the index improves the performance in your case?

**Comment by nijel** [comment]
Recent I've made some improvements in configuring similarity in 4fb791461e9f7350b07d1d5b044d6c9b96574305, b45251b6b575c7cb1dabf30f003b99767d5452b4 and d09738ccf0923046ba06f067b4784a4dec06911d. There still might be a room for improvement, but let's close this for now.
